server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseBackgroundSyncAction.java

```diff
@@ -119,19 +120,21 @@ public void backgroundSync(
     }
 
     @Override
-    protected PrimaryResult<Request, ReplicationResponse> shardOperationOnPrimary(final Request request, final IndexShard primary) {
+    protected PrimaryResult<Request, ReplicationResponse> shardOperationOnPrimary(
+            final Request request,
+            final IndexShard primary) throws WriteStateException {
         Objects.requireNonNull(request);
         Objects.requireNonNull(primary);
-        primary.afterWriteOperation();
+        primary.persistRetentionLeases();
         return new PrimaryResult<>(request, new ReplicationResponse());
     }
 
     @Override
-    protected ReplicaResult shardOperationOnReplica(final Request request, final IndexShard replica){
+    protected ReplicaResult shardOperationOnReplica(final Request request, final IndexShard replica) throws WriteStateException {
```

Comment:
I think we can remove RetentionLeaseBackgroundSyncAction since it is similar to RetentionLeaseSyncAction. We can do it in a follow-up.

Reply:
I pushed 1675507ae8f20e517303db1aed9f5fd02993251d.

Reply:
Note that one is a write action and the other is only a replication action. This is a deliberate difference.

Reply:
ah, remember it now. Thanks @jasontedor.